src/sqlfluff/dialects/dialect_tsql.py

```diff
@@ -0,0 +1,177 @@
+# common and match the largest areas and so are sufficient.
+tsql_dialect.sets("bracket_pairs").update(
+    [
+        ("round", "StartBracketSegment", "EndBracketSegment", True),
+        ("square", "StartSquareBracketSegment", "EndSquareBracketSegment", True),
+        ("curly", "StartCurlyBracketSegment", "EndCurlyBracketSegment", False),
+    ]
+)
+
+
+# Update only RESERVED Keywords
+# tsql_dialect.sets("reserved_keywords").clear()
+# tsql_dialect.sets("reserved_keywords").update(RESERVED_KEYWORDS)
+
+tsql_dialect.add(
+    AtSignLiteralSegment=NamedParser(
+        "atsign",
+        CodeSegment,
+        name="atsign_literal",
+        type="literal",
+        trim_chars=("@",),
+    ),
+)
+
+tsql_dialect.add(    
+    AtSignSignSegment=StringParser(
+        "@", SymbolSegment, name="atsign", type="user_designator"
+    ),   
+    SessionVariableNameSegment=RegexParser(
+        r"[@][a-zA-Z0-9_]*",
+        CodeSegment,
+        name="declared_variable",
+        type="variable",
+    ),
+)
+
+tsql_dialect.replace(
+    ParameterNameSegment=RegexParser(
+        r"[@][A-Za-z0-9_]+", CodeSegment, name="parameter", type="parameter"
+    )
+)
+
+tsql_dialect.insert_lexer_matchers(
+    [
+        RegexLexer(
+            "atsign",
+            r"[@][a-zA-Z0-9_]+",
+            CodeSegment,
+        ),
+    ],
+    before="code",
+)
+
+
+@tsql_dialect.segment(replace=True)
+class CreateFunctionStatementSegment(BaseSegment):
+    """A `CREATE FUNCTION` statement.
+
+    This version in the ANSI dialect should be a "common subset" of the
+    structure of the code for those dialects.
+    postgres: https://www.postgresql.org/docs/9.1/sql-createfunction.html
+    snowflake: https://docs.snowflake.com/en/sql-reference/sql/create-function.html
+    bigquery: https://cloud.google.com/bigquery/docs/reference/standard-sql/user-defined-functions
+    """
+
+    type = "create_function_statement"
+
+    match_grammar = Sequence(
+        "CREATE",
+        Sequence("OR", "ALTER", optional=True),
+        Ref("TemporaryGrammar", optional=True),
+        "FUNCTION",
+        Anything(),
+    )
+    parse_grammar = Sequence(
+        "CREATE",
+        Sequence("OR", "ALTER", optional=True),
+        Ref("TemporaryGrammar", optional=True),
+        "FUNCTION",       
+        Sequence(
+            Ref("StartSquareBracketSegment", optional=True),
+            Ref("FunctionNameSegment"),
+            Ref("EndSquareBracketSegment", optional=True),
+            Ref("DotSegment"), optional=True),               
+        Sequence(
+            Ref("StartSquareBracketSegment", optional=True),
+            Ref("FunctionNameSegment"),
+            Ref("EndSquareBracketSegment", optional=True)),
+        Ref("FunctionParameterListGrammar"),
+        Sequence(  # Optional function return type
+            "RETURNS",
+            Ref("DatatypeSegment"),
+            optional=True,
+        ),
+        Sequence("AS"),
```

Comment:
Shouldn't AS, BEGIN, and END just use `Ref`, not `Sequence`? Maybe GO as well. (I can't remember offhand if `Ref` allows the `optional` keyword or not.

Reply:
I couldn't get `OptionallyBracketed` to work. It doesn't take a `bracket_type` argument so guess it's hardcoded to `(` and `)`? Probably potential to add add this optional param but maybe out of scope for this PR?

And for raw words you don't even need `Ref` - just the plain `"AS"`. Already told @ericmuijs this offline.

`Ref` does work for `GO` though to added the `optional=True` part and you're right it probably is better to use that than the `Sequence` there.